### PR TITLE
Improve brokerage model python wrapper

### LINKED ISSUE #7653: Improve BrokerageModelPythonWrapper
#### Expected Behavior

- BrokerageModelPythonWrapper can return a python based model
#### Actual Behavior

- BrokerageModelPythonWrapper just directly returns what the underlying model returns
#### Potential Solution

- BrokerageModelPythonWrapper implementation should be smarter and detect that the response he was given from the underlying python BrokerageModel is a python object and should be wrapped with the appropriate python wrapper 
#### Reproducing the Problem

```csharp
        [Test]
        public void Bug()
        {
            using (Py.GIL())
            {
                dynamic PyCustomBrokerageModel = PyModule.FromString("testModule",
                    @$"
from AlgorithmImports import *

class CustomFillModel(ImmediateFillModel):
    def __init__(self):
        super().__init__()

    def MarketFill(self, asset, order):
        raise ValueError(""Pepe"")        

class CustomBrokerageModel(DefaultBrokerageModel):
    def GetFillModel(self, security):
        return CustomFillModel()
                ").GetAttr("CustomBrokerageModel");

                var security = GetSecurity(Symbols.SPY);
                var model = new BrokerageModelPythonWrapper(PyCustomBrokerageModel());
                var result = model.GetFillModel(security);
            }
        }
```
#### System Information

N/A
#### Checklist

- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

- [x] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by Marinovsky**
#### Description

So far, `BrokerageModelPythonWrapper.cs` had a bug when returning Python models, such as Benchmark, FeeModel, FillModel, etc. Instead of returning the python model, it returned the underlying C# model. Now, that problem has been fixed and unit tests has been added to cover the changes.
#### Related Issue

Closes #7653
Partially https://github.com/QuantConnect/Lean/issues/7676
#### Motivation and Context

With this change, user's algorithms will work with the expected Python custom models when using a custom brokerage model
#### Requires Documentation Change

N/A
#### How Has This Been Tested?

To test that the methods from `BrokerageModelPythonWrapper` actually returned the custom python model, there were created unit tests that assert the custom implementation was called instead of the underlying C# implementation. Additionally, two regression tests were made to assert the `SettlementModelPythonWrapper` and `BenchmarkPythonWrapper` methods worked as expected.
#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Common/Python/BrokerageModelPythonWrapper.cs [hunk 1]
```diff
@@ -221,7 +232,12 @@ public IFillModel GetFillModel(Security security)
         {
             using (Py.GIL())
             {
-                return (_model.GetFillModel(security) as PyObject).GetAndDispose<IFillModel>();
+                var fillModel = _model.GetFillModel(security) as PyObject;
+                if (fillModel.TryConvert<IFillModel>(out var csharpFillModel))
+                {
+                    return csharpFillModel;
+                }
+                return (new FillModelPythonWrapper(fillModel));
```
Comment (jhonabreul) [inline Common/Python/BrokerageModelPythonWrapper.cs:240]: nit - parentheses

File: Common/Python/BrokerageModelPythonWrapper.cs [hunk 2]
```diff
@@ -262,8 +283,7 @@ public ISettlementModel GetSettlementModel(Security security, AccountType accoun
         {
             using (Py.GIL())
             {
-                return (_model.GetSettlementModel(security, accountType)
-                    as PyObject).GetAndDispose<ISettlementModel>();
+                return (_model.GetSettlementModel(security, accountType) as PyObject).GetAndDispose<ISettlementModel>();
```
Comment (jhonabreul) [inline Common/Python/BrokerageModelPythonWrapper.cs:286]: Does this need the same `TryConvert`... like the other methods?

File: Common/Python/BrokerageModelPythonWrapper.cs [hunk 3]
```diff
@@ -320,8 +350,7 @@ public IBuyingPowerModel GetBuyingPowerModel(Security security, AccountType acco
         {
             using (Py.GIL())
             {
-                return (_model.GetBuyingPowerModel(security, accountType)
-                    as PyObject).GetAndDispose<IBuyingPowerModel>();
+                return (_model.GetBuyingPowerModel(security, accountType) as PyObject).GetAndDispose<IBuyingPowerModel>();
```
Comment (jhonabreul) [inline Common/Python/BrokerageModelPythonWrapper.cs:353]: Same comment as for the settlement model

File: Tests/Python/PythonWrapperTests.cs [hunk 4]
```diff
@@ -59,6 +60,80 @@ public void DoesNotThrowWhenDerivedFromCSharpModel()
...
+                    {"Treynor Ratio", "-3.226"},
+                    {"Total Fees", "$0.00"},
+                    {"Estimated Strategy Capacity", "$0"},
+                    {"Lowest Capacity Asset", ""},
+                    {"Portfolio Turnover", "0%"},
+                    {"OrderListHash", "d41d8cd98f00b204e9800998ecf8427e"}
+                },
+                Language.Python,
+                AlgorithmStatus.Completed,
+                algorithmLocation: "../../../Algorithm.Python/CustomSettlementModelRegressionAlgorithm.py");
```
Comment (jhonabreul) [inline Tests/Python/PythonWrapperTests.cs:97]: nit - additional tab missing

File: Tests/Python/PythonWrapperTests.cs [hunk 5]
```diff
@@ -59,6 +60,80 @@ public void DoesNotThrowWhenDerivedFromCSharpModel()
...
+            }
+
+            [Test]
+            public void BenchmarkModelPythonWrapperWorks()
+            {
+                var results = AlgorithmRunner.RunLocalBacktest("CustomBenchmarkRegressionAlgorithm",
+                new Dictionary<string, string>()
```
Comment (jhonabreul) [inline Tests/Python/PythonWrapperTests.cs:104]: nit- additional tab missing

File: Tests/Common/Brokerages/BrokerageModelTests.cs [hunk 6]
```diff
@@ -214,6 +223,418 @@ public void GetsCorrectBuyingPowerModelForSecurityAndAccountType(IBrokerageModel
...
+class CustomBrokerageModel(DefaultBrokerageModel):
+    def GetFillModel(self, security):
+        return CustomFillModel()
+                ").GetAttr("CustomBrokerageModel");
+
+                var security = GetSecurity(Symbols.SPY);
+                var model = new BrokerageModelPythonWrapper(PyCustomBrokerageModel());
+                var fillModel = model.GetFillModel(security);
+                Assert.AreEqual(typeof(FillModelPythonWrapper), fillModel.GetType());
+                Assert.Throws<PythonException>(() => ((dynamic)fillModel).MarketFill(security, new Mock<MarketOrder>().Object));
```
Comment (jhonabreul) [inline Tests/Common/Brokerages/BrokerageModelTests.cs:251]: Could also assert that it is a `ValueError` exception and its message content

File: Tests/Common/Brokerages/BrokerageModelTests.cs [hunk 7]
```diff
@@ -214,6 +223,418 @@ public void GetsCorrectBuyingPowerModelForSecurityAndAccountType(IBrokerageModel
...
+class CustomBrokerageModel(DefaultBrokerageModel):
+    def GetBenchmark(self, securities):
+        return CustomBenchmarkModel()
+                ").GetAttr("CustomBrokerageModel");
+                var timeKeeper = new TimeKeeper(DateTime.Now);
+                var securityManager = new SecurityManager(timeKeeper);
+                var model = new BrokerageModelPythonWrapper(PyCustomBrokerageModel());
+                var benchmarkModel = model.GetBenchmark(securityManager);
+                Assert.AreEqual(typeof(BenchmarkPythonWrapper), benchmarkModel.GetType());
+                Assert.Throws<PythonException>(() => ((dynamic)benchmarkModel).Evaluate(DateTime.Now));
```
Comment (jhonabreul) [inline Tests/Common/Brokerages/BrokerageModelTests.cs:308]: Same comment as for the fill model test

File: Tests/Common/Brokerages/BrokerageModelTests.cs [hunk 8]
```diff
@@ -214,6 +223,418 @@ public void GetsCorrectBuyingPowerModelForSecurityAndAccountType(IBrokerageModel
...
+    def GetFeeModel(self, securities):
+        return CustomFeeModel()
+                ").GetAttr("CustomBrokerageModel");
+                var security = GetSecurity(Symbols.SPY);
+                var model = new BrokerageModelPythonWrapper(PyCustomBrokerageModel());
+                var feeModel = model.GetFeeModel(security);
+                Assert.AreEqual(typeof(FeeModelPythonWrapper), feeModel.GetType());
+                var order = new Mock<Order>();
+                var orderParameters = new Mock<OrderFeeParameters>(security, order.Object);
+                Assert.Throws<PythonException>(() => ((dynamic)feeModel).GetOrderFee(orderParameters.Object));
```
Comment (jhonabreul) [inline Tests/Common/Brokerages/BrokerageModelTests.cs:363]: Same comment as for the fill model test

File: Tests/Common/Brokerages/BrokerageModelTests.cs [hunk 9]
```diff
@@ -214,6 +223,418 @@ public void GetsCorrectBuyingPowerModelForSecurityAndAccountType(IBrokerageModel
...
+                var model = new BrokerageModelPythonWrapper(PyCustomBrokerageModel());
+                var settlementModel = model.GetSettlementModel(security);
+                Assert.AreEqual(typeof(SettlementModelPythonWrapper), settlementModel.GetType());
+                var algorithm = new AlgorithmStub();
+                algorithm.SetDateTime(DateTime.Now);
+                var portfolio = algorithm.Portfolio;
+                var appyFundsParameters = new ApplyFundsSettlementModelParameters(portfolio, security, DateTime.Now, new CashAmount(1000, Currencies.USD), null);
+                Assert.Throws<PythonException>(() => ((dynamic)settlementModel).ApplyFunds(appyFundsParameters));
+                var scanParameters = new ScanSettlementModelParameters(portfolio, security, DateTime.UtcNow);
+                Assert.Throws<PythonException>(() => ((dynamic)settlementModel).Scan(scanParameters));
```
Comment (jhonabreul) [inline Tests/Common/Brokerages/BrokerageModelTests.cs:421]: Same comment as for the fill model test

File: Tests/Common/Brokerages/BrokerageModelTests.cs [hunk 10]
```diff
@@ -214,6 +223,418 @@ public void GetsCorrectBuyingPowerModelForSecurityAndAccountType(IBrokerageModel
...
+class CustomBrokerageModel(DefaultBrokerageModel):
+    def GetSlippageModel(self, security):
+        return CustomSlippageModel()
+                ").GetAttr("CustomBrokerageModel");
+                var security = GetSecurity(Symbols.SPY);
+                var model = new BrokerageModelPythonWrapper(PyCustomBrokerageModel());
+                var slippageModel = model.GetSlippageModel(security);
+                Assert.AreEqual(typeof(SlippageModelPythonWrapper), slippageModel.GetType());
+                var order = new Mock<Order>();
+                Assert.Throws<PythonException>(() => ((dynamic)slippageModel).GetSlippageApproximation(security, order.Object));
```
Comment (jhonabreul) [inline Tests/Common/Brokerages/BrokerageModelTests.cs:473]: Same comment as for the fill model test

File: Tests/Common/Brokerages/BrokerageModelTests.cs [hunk 11]
```diff
@@ -214,6 +223,418 @@ public void GetsCorrectBuyingPowerModelForSecurityAndAccountType(IBrokerageModel
...
+
+class CustomBrokerageModel(DefaultBrokerageModel):
+    def GetBuyingPowerModel(self, security):
+        return CustomBuyingPowerModel()
+                ").GetAttr("CustomBrokerageModel");
+                var security = GetSecurity(Symbols.SPY);
+                var model = new BrokerageModelPythonWrapper(PyCustomBrokerageModel());
+                var buyingPowerModel = model.GetBuyingPowerModel(security);
+                Assert.AreEqual(typeof(BuyingPowerModelPythonWrapper), buyingPowerModel.GetType());
+                Assert.Throws<PythonException>(() => ((dynamic)buyingPowerModel).GetLeverage(security));
```
Comment (jhonabreul) [inline Tests/Common/Brokerages/BrokerageModelTests.cs:520]: Same comment as for the fill model test

File: Tests/Common/Brokerages/BrokerageModelTests.cs [hunk 12]
```diff
@@ -214,6 +223,418 @@ public void GetsCorrectBuyingPowerModelForSecurityAndAccountType(IBrokerageModel
...
+
+class CustomBrokerageModel(DefaultBrokerageModel):
+    def GetShortableProvider(self, security):
+        return CustomShortableProvider()
+                ").GetAttr("CustomBrokerageModel");
+                var security = GetSecurity(Symbols.SPY);
+                var model = new BrokerageModelPythonWrapper(PyCustomBrokerageModel());
+                var shortableProvider = model.GetShortableProvider(security);
+                Assert.AreEqual(typeof(ShortableProviderPythonWrapper), shortableProvider.GetType());
+                Assert.Throws<PythonException>(() => ((dynamic)shortableProvider).ShortableQuantity(security.Symbol, DateTime.Now));
```
Comment (jhonabreul) [inline Tests/Common/Brokerages/BrokerageModelTests.cs:566]: Same comment as for the fill model test

File: Tests/Common/Brokerages/BrokerageModelTests.cs [hunk 13]
```diff
@@ -214,6 +223,418 @@ public void GetsCorrectBuyingPowerModelForSecurityAndAccountType(IBrokerageModel
...
+class CustomBrokerageModel(DefaultBrokerageModel):
+    def GetMarginInterestRateModel(self, security):
+        return CustomMarginInterestRateModel()
+                ").GetAttr("CustomBrokerageModel");
+                var security = GetSecurity(Symbols.SPY);
+                var model = new BrokerageModelPythonWrapper(PyCustomBrokerageModel());
+                var marginInterestRateModel = model.GetMarginInterestRateModel(security);
+                Assert.AreEqual(typeof(MarginInterestRateModelPythonWrapper), marginInterestRateModel.GetType());
+                var parameters = new MarginInterestRateParameters(security, DateTime.Now);
+                Assert.Throws<PythonException>(() => ((dynamic)marginInterestRateModel).ApplyMarginInterestRate(parameters));
```
Comment (jhonabreul) [inline Tests/Common/Brokerages/BrokerageModelTests.cs:613]: Same comment as for the fill model test

**Review by jhonabreul — CHANGES_REQUESTED** [review]
Thanl you! Leaving some minor comments

File: Common/Python/BrokerageModelPythonWrapper.cs (hunk 2 above)
Comment (Marinovsky) [reply Common/Python/BrokerageModelPythonWrapper.cs:286]: Yeah, but since that method has been deprecated I didn't know if I should also modify it 🤔

**Review by jhonabreul — APPROVED** [review]
Thank you!

**Review by Martin-Molinero — APPROVED** [review]
Nice work 💪, thank you!

**Pull request merged**